### ISSUE TITLE
Android 12 + Kernel 5.4, but capture tls return all 0

### ISSUE THREAD

**Issue by HorseLuke** [gojue/ecapture#296]
**Describe the bug**
I have a android device with Android 12 + Kernel 5.4, but capture tls return all 0.
Device is Moto g71s (XT2225-2)

**To Reproduce**

```
rhodep:/system/bin/ecapture-nocore # zcat /proc/config.gz |grep CONFIG_DEBUG_INFO_BTF
# CONFIG_DEBUG_INFO_BTF is not set

rhodep:/system/bin/ecapture-nocore # ./ecapture -v
ecapture version:       androidgki_aarch64:0.4.11-20230107-7b66305:5.4.0-104-generic

rhodep:/system/bin/ecapture-nocore # ./ecapture tls

```

**Expected behavior**

caputre tls return cleartext

**Screenshots**

```
rhodep:/system/bin/ecapture-nocore # ./ecapture tls
tls_2023/01/17 16:14:41 ECAPTURE :: ecapture Version : androidgki_aarch64:0.4.11-20230107-7b66305:5.4.0-104-generic
tls_2023/01/17 16:14:41 ECAPTURE :: Pid Info : 662
tls_2023/01/17 16:14:41 ECAPTURE :: Kernel Info : 5.4.147
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        module initialization
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        master key keylogger: ecapture_masterkey.log
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        Module.Run()
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        UPROBE MODEL
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        OpenSSL/BoringSSL version not found, used default version :android_default
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        HOOK type:2, binrayPath:/apex/com.android.conscrypt/lib64/libssl.so
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        Hook masterKey function:SSL_in_init
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        target all process.
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        target all users.
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        BPF bytecode filename:user/bytecode/boringssl_1_1_1_kern.o
tls_2023/01/17 16:14:41 EBPFProbeOPENSSL        module started successfully.
tls_2023/01/17 16:14:41 ECAPTURE ::     start 1 modules
tls_2023/01/17 16:14:45 UUID:987_1244_pool-5-thread-1_0_1, Name:DefaultParser, Type:0, Length:1972
tls_2023/01/17 16:14:45
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000040  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000060  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000070  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000080  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000090  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000d0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000e0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000100  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000110  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000120  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000140  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000150  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000160  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000170  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000180  00 00 00 00 00 00 00 00  00 00 00 00 00           |.............|
```

**Linux Server/Android (please complete the following information):**

- Env: [run `make env` to get the environment variables]
- OS: Android 12
- Arch: aarch64
- Kernel Version: 5.4.147-moto
- Version: v0.4.11 NOCORE

**Additional context**
I read some issues  (link: https://github.com/gojue/ecapture/issues/293#issuecomment-1376866789 ), is that kernel 5.4 is not supported on aarch64?

**Comment by cfc4n** [comment]
> I read some issues (link: https://github.com/gojue/ecapture/issues/293#issuecomment-1376866789 ), is that kernel 5.4 is not supported on aarch64?

you are right.  

incorret , Linux kernel Arm64(aarch64) supported this feature with `bpf_probe_read_user` at  https://cdn.kernel.org/pub/linux/kernel/v5.x/ChangeLog-5.5

so, you need used Android (linux) kernel 5.5 or newer to used eCapture on `arm64(aarch64)` .

**Comment by HorseLuke** [comment]
OK, thanks

**Comment by HorseLuke** [comment]
With a lot of effort, I managed to run ecapture on android 11/12 with kernel 5.4 aarch64. Here is a brief step, tested on Debian 11 with docker.

[Easy version]

Buy a new phone with kernel 5.10 or higher. :)

[Normal version]

1. Grab kernel 5.4 source codes
  1.1. If you can, Grab kernel 5.4 source codes from mobile manufacturer.
  1.2 If not, detect whether mobile is compatible with GKI 1.0 [2]. If so, grab kernel 5.4 source codes from aosp.

    (A) Grab stock boot.img.

    (B) use "tools/mkbootimg/unpack_bootimg.py --boot_img=boot.img"[1] to detect os_version. 11 means android11-5.4 branch, 12 means android12-5.4 branch.

    (C) Download the GKI boot image from ci.android.com. android11-5.4 branch use  https://ci.android.com/builds/branches/aosp-android11-gsi/grid ,  android12-5.4 branch is https://ci.android.com/builds/branches/aosp-android12-gsi/grid . Download aosp_arm64-img-xxxx.zip from 	aosp_arm64 target, and extract to pick file boot-5.4.img

    (D) boot mobile into fastboot mode, connect to pc, then run "fastboot.exe boot boot-5.4.img". If success, it will boot successfully, and kernel version will change in about setting; If not, it will boot failed, or kernel version remains unchanged in about setting, you have to buy a new phone with kernel 5.10 or higher :( .

    (E) If success in D, go to grab kernel 5.4 source codes from aosp,  android11-5.4 branch use common-android11-5.4 , android12-5.4 branch use common-android12-5.4 branch[3]
		
2. Patch probe_read_{user,kernel}() and other codes from kernel upstream

	Patch ALL CODES from https://github.com/torvalds/linux/commit/358fdb456288d48874d44a064a82bfb0d9963fa0
	
3. Rebuild patched kernel and generate boot.img

    CONFIG_DEBUG_INFO_BTF can be Y or not set. 
    Tips: Some phone may not boot or bootloop with CONFIG_DEBUG_INFO_BTF=y, guess that rebuilded kernel with BTF info is too large to boot.
	
4. Generate BTF
    If CONFIG_DEBUG_INFO_BTF  is Y in step 3, ignore this step. If not, do this step
    After rebuild, you will get vmlinux in output dir. use this command to generate btf

```
cp -a vmlinux vmlinux.btf
pahole -J vmlinux.btf
```

5. Generate vmlinux.h

Use  bpftool to generate vmlinux.h, it will take very long time. Run it outside docker container and newest bpftool version may take less time.

```
bpftool btf dump file vmlinux.btf  format c > vmlinux.h
```
	
6. Rebuild ecapture with generated vmlinux.h and no core mode.
	You can run rebuild process in docker arm64 container on x64 host[4][5]. This will save lots of time.
	Before run rebuild, modify Makefile, hardcode UNAME_M and UNAME_R like:

```
UNAME_M := aarch64
UNAME_R := 5.4.226-android11-2
```	
Then rebuild it with command like these

```
ANDROID=1 \
	KERN_BUILD_PATH=/aosp1/android-kernel/out/android11-5.4/common \
	KERN_SRC_PATH=/aosp1/android-kernel/common \
	make nocore
```
	
Do not forget modify KERN_BUILD_PATH and KERN_SRC_PATH
	
	
7. Now bin/ecapture can be run on patched kernel 5.4.



------

[1]https://android.googlesource.com/platform/system/tools/mkbootimg/+/master/unpack_bootimg.py
[2] https://source.android.com/docs/core/architecture/kernel/generic-kernel-image
[3] https://source.android.com/docs/setup/build/building-kernels
[4] https://docs.docker.com/build/building/multi-platform/
[5] https://devopstales.github.io/linux/running_and_building_multi_arch_containers/

**Comment by HorseLuke** [comment]
Open discussion: https://github.com/gojue/ecapture/discussions/308
